auto pagination

amarjandu opened the pull request:
Allows for auto pagination for hca commands, turns on this feature by default.
Adds support for optional `--no-paginate` to fall back to single page response. 
Closes #461

kislyuk commented on hca/util/__init__.py:
```diff
@@ -198,7 +199,29 @@ def iterate(self, **kwargs):
-            yield page.json()
+            data = page.json()
+            content_key = page.headers.get("X-OpenAPI-Paginated-Content-Key", "results")
+            data['_internal_api_content_key'] = content_key
```
I would prefer it if we could find a way not to pass this in-band.

amarjandu replied:
I think an alternative way to perform this would be to populate the header value in the `swagger.json` page, then create a internal function that can extract the value from 
`_PaginatingClientMethodFactory.method_data`. This would require changes to the DSS Swagger, and potentially Fusillade/Query Service Swagger pages as well.

kislyuk replied:
You can use `_PaginatingClientMethodFactory. _get_raw_pages()`, the method `paginate()` wraps. `_get_raw_pages()` yields entire responses. You can read the paginated content key from the response headers.

kislyuk commented on hca/util/__init__.py:
```diff
@@ -198,7 +199,29 @@ def iterate(self, **kwargs):
+            yield data
+
+    def _cli_call(self, cli_args):
+        if str.lower(vars(cli_args)['page']) == 'false':
```
Why not just use `cli_args.page` here?

amarjandu replied:
:) thanks for catching that.

kislyuk commented on hca/util/__init__.py:
```diff
@@ -514,11 +537,14 @@ def _build_client_method(self, http_method, http_path, method_data):
 
+        if method_supports_pagination:
+            parameters['page'] = dict(description='Use the internal CLI paging mechanism', required=False,
+                                      name='page', default="True", type=str)
```
Consider using this instead:

add_argument("--no-paginate", action="store_false", dest="paginate")

This will allow you to say `hca dss post-search --no-paginate ...` and the argument to consume will be `args.paginate`, a boolean.

amarjandu replied:
Much cleaner, thanks!

amarjandu commented:
> This needs a test with either a small page size or a mock. I can help with a mock if needed.

I was having a little bit of trouble with the MOCK functionality. I'm having a hard time figuring out where is an effective place to put the mock.  I don't think a small page size would work as `dss-dev` is cleared out every day, which can cause inconsistent/incomplete testing. 

Please let me know if you think the approach to getting the `X-OpenAPI-Paginated-Content-Key` from the `swagger.json` is a better approach.

kislyuk commented on hca/util/__init__.py:
```diff
@@ -198,7 +199,32 @@ def iterate(self, **kwargs):
-            yield page.json()
+            data = page.json()
+            content_key = page.headers.get("X-OpenAPI-Paginated-Content-Key", "results")
+            yield data, content_key
```
Please don't change the signature of `paginate()`.

kislyuk commented:
@amarjandu Here is a starting point for a test case that I would use to unit test this with a mock:
```python
import json, io, unittest, unittest.mock
import requests
from requests.models import Response
import hca


class SessionMock:
    status_code = requests.codes.ok

    def request(self, *args, **kwargs):
        payload = {"results": ["foo", "bar"]}
        res = Response()
        res.status_code = self.status_code
        res.headers = {"foo": "bar"}
        res.raw = io.BytesIO(json.dumps(payload).encode())
        return res


class MockTestCase(unittest.TestCase):
    def test_with_mock(self):
        hca.util.SwaggerClient.get_session = SessionMock
        for page in hca.dss.DSSClient().get_bundles_all.paginate(replica="aws"):
            print(page)


if __name__ == "__main__":
    unittest.main()
```
You can do fancier things with `unittest.mock`, but I started simple here and I think we can do all we need here with this approach.

Regarding the paginated content key: please see my other comment about using `_PaginatingClientMethodFactory._get_raw_pages()` in the caller.

amarjandu commented:
@kislyuk

 I made some changes to this PR. I removed the auto-pagination features from the python bindings, this forces a developer who is using the bindings to go through the `_PaginatingClientMethodFactory.paginate` method, if they want to flex the paging feature. However the when using the cli-commands auto-pagination will be enabled by default.

This could be seen as 'unexpected' behavior, perhaps we should turn off auto-paging in the cli-commands by default, and change the argument to `--paginate`.

I implemented a version of testing with your suggestion, but it is erroring in testing within allspark/travis, when i test locally on my machine the added tests pass. It seems that it is expecting more attributes within the SessionMock to be present, could we change where we are performing the Mock?

kislyuk commented:
Actually I think we *should* paginate and fetch all results by default in the CLI. I think that's the least surprising behavior.

The reason you're seeing the error is that the test is trying to call `.get()` on the session mock to fetch the Swagger API definition; it's succeeding locally because a copy of the definition has already been cached.

To fix this, define the `get()` method on the mock:
```
class SessionMock:
    ...
    def get(self, *args, **kwargs):
        kwargs["method"] = "GET"
        return self.request(*args, **kwargs)
```

kislyuk commented on hca/util/__init__.py:
```diff
@@ -200,6 +201,31 @@ def paginate(self, **kwargs):
+
+    def __call__(self, client, **kwargs):
+        # auto pagination should not be available for python bindings, use paginate attribute instead.
+        return super().__call__(client, **kwargs)
```
This `__call__()` declaration seems to be a no-op. Can you remove it?

amarjandu replied:
Resolved.
